### PR TITLE
Convert runInteraction to async/await

### PR DESCRIPTION
**Pull request by clokep**
This converts `runInteraction` to async/await.

This should be the last "main" part of #7988.

### CODE REVIEW

**Comment by clokep** [comment]
This has a conflict already so I'll likely rebase this after #8140 is merged.

**Comment by clokep** [comment]
@erikjohnston previously stated that we might consider it an anti-pattern to return `Awaitable` instead of awaiting on it. I don't know if we want to "fix" that as part of this work or leave it to a future piece of work? We could do this module-by-module or so if we wanted.

**Comment by clokep** [comment]
I'm putting review on this, but it's much larger than I'd like. Thoughts / ideas of how to slim it down and make it more reviewable would be appreciated!

**Comment by anoadragon453** [comment]
While it would roughly 2x the size of this PR, it's not too much more of a mental exercise to review converting to Await + converting the method itself to async and adding an `await` to the `return` statement.

If you did that and then proposed review by modules I think that could be managed.

**Comment by clokep** [comment]
I thought a bit more about this and I *think* it will make sense to do this method by method, e.g. make `simple_select_one` into an async function and chase down all the users of that. Then after all those are done, do `runInteraction` directly (which should essentially just be this PR, but with a bunch of stuff removed).

**Comment by clokep** [comment]
There's a few PRs that were spun out of this. They don't really depend on this, but it would be nice to merge them first so that the comments and type hints are all correct in referring to awaitables.

File: synapse/storage/database.py
Comment (clokep) [inline synapse/storage/database.py:511]: I think this can be:

```suggestion
    async def runInteraction(
        self, desc: str, func: "Callable[..., R]", *args: Any, **kwargs: Any
    ) -> R:
```

But mypy then complains:

> synapse/storage/database.py:550: error: Incompatible return value type (got "R", expected "R")  [return-value]

I suspect something isn't getting piped through `runWithConnection` / `new_transaction` properly?

Comment (ilmari) [reply synapse/storage/database.py:511]: Does the `inner_func` in `runWithConnection` need annotating as returning `R`?

Comment (erikjohnston) [reply synapse/storage/database.py:511]: So I think the disconnect is happening because mypy doesn't know that the return type of `self.new_transaction` is the same `R` as returned by `func`, due to the indirection of `runWithConnection`.

The signature `runWithConnection(func, *args, **kwargs)` means that when we call it with `runWithConnection(self.new_transaction, func)` mypy only sees us calling `self.new_transaction(*args)` as the typing of `func` is lost due to the variadic parameters.

Changing it to `runWithConnection(lambda conn: self.new_transaction(conn, func))` works, but it may be simpler to do a `result = cast(R, result)` instead.

Comment (clokep) [reply synapse/storage/database.py]: Woot! Thanks. See ca31cfeb5572d3abda2790ccdc9995b79502dd0a for this.

**Review by erikjohnston — APPROVED** [review]

**Pull request merged**
